compiler/rustc_mir_transform/src/dest_prop.rs

```diff
@@ -238,6 +239,12 @@ impl<'tcx> MirPass<'tcx> for DestinationPropagation {
             apply_merges(body, tcx, &merges, &merged_locals);
         }
 
+        if round_count != 0 {
+            // Merging can introduce ill-formed overlap between moved arguments and/or call
+            // destination in an unreachable code.
+            remove_dead_blocks(tcx, body);
```

Comment:
What is the structural reason that this can only happen in dead code?

Reply:
The logic that prevents us from introducing this kind of UB is [here](https://github.com/rust-lang/rust/blob/7632db0e87d8adccc9a83a47795c9411b1455855/compiler/rustc_mir_transform/src/dest_prop.rs#L454). That loop only iterates over reachable blocks, which means we can accidentally introduce these kinds of ICEs in unreachable blocks.